base/sort.jl

```diff
@@ -529,75 +530,107 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
-        end
+# TODO replace with rand(lo:hi) if Random is ever a part of Base
+function select_pivot(v::AbstractVector, lo::Integer, hi::Integer, o::Ordering)
+    mi = midpoint(lo, hi)
+    @inbounds a, b, c = v[lo], v[mi], v[hi]
+
+    if lt(o, b, a)
+        b, a = a, b
+        lo, mi = mi, lo
+    end
 
-        if lt(o, v[hi], v[lo])
-            if lt(o, v[hi], v[mi])
-                v[hi], v[lo], v[mi] = v[lo], v[mi], v[hi]
-            else
-                v[hi], v[lo] = v[lo], v[hi]
-            end
-        end
+    # a is stably less than b
 
-        # return the pivot
-        return v[lo]
+    if lt(o, c, a)
+        lo
+    elseif lt(o, c, b)
+        hi
+    else
+        mi
     end
 end
 
-# partition!
+# select a pivot, partition v[lo:hi] according
+# to the pivot, and store the result in t[lo:hi].
 #
-# select a pivot, and partition v according to the pivot
-
-function partition!(v::AbstractVector, lo::Integer, hi::Integer, o::Ordering)
-    pivot = selectpivot!(v, lo, hi, o)
-    # pivot == v[lo], v[hi] > pivot
-    i, j = lo, hi
-    @inbounds while true
-        i += 1; j -= 1
-        while lt(o, v[i], pivot); i += 1; end;
-        while lt(o, pivot, v[j]); j -= 1; end;
-        i >= j && break
-        v[i], v[j] = v[j], v[i]
+# returns (pivot, pivot_index) where pivot_index is the location the pivot
+# should end up, but does not set t[pivot_index] = pivot
+function partition!(t::AbstractVector, lo::Integer, hi::Integer, o::Ordering, v::AbstractVector, rev::Bool)
+    pivot_index = select_pivot(v, lo, hi, o)
+    trues = 0
+    @inbounds begin
+        pivot = v[pivot_index]
+        while lo < pivot_index
+            x = v[lo]
+            fx = rev ? !lt(o, x, pivot) : lt(o, pivot, x)
+            t[(fx ? hi : lo) - trues] = x
+            trues += fx
+            lo += 1
+        end
+        while lo < hi
+            x = v[lo+1]
+            fx = rev ? lt(o, pivot, x) : !lt(o, x, pivot)
+            t[(fx ? hi : lo) - trues] = x
+            trues += fx
+            lo += 1
+        end
     end
-    v[j], v[lo] = pivot, v[j]
 
-    # v[j] == pivot
-    # v[k] >= pivot for k > j
-    # v[i] <= pivot for i < j
-    return j
+    # pivot_index = lo-trues
+    # t[pivot_index] is whatever it was before
+    # t[<pivot_index] <* pivot, stable
+    # t[>pivot_index] >* pivot, reverse stable
+
+    pivot, lo-trues
 end
 
-function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::QuickSortAlg, o::Ordering)
-    @inbounds while lo < hi
-        hi-lo <= SMALL_THRESHOLD && return sort!(v, lo, hi, SMALL_ALGORITHM, o)
-        j = partition!(v, lo, hi, o)
-        if j-lo < hi-j
-            # recurse on the smaller chunk
-            # this is necessary to preserve O(log(n))
-            # stack space in the worst case (rather than O(n))
-            lo < (j-1) && sort!(v, lo, j-1, a, o)
+# Needed for bootstrapping because `view` is not available.
+function reverse_view!(v, lo, hi)
```

Comment:
Isn't this the same as `reverse!` from `Base`?

Reply:
Yes, it is, but we can't use `view` in sort.jl because sorting comes before views in bootstrapping. I've added an additional comment to clarify, and tests to verify that it is true.

Reply:
Thanks for the comment, but I don't see the difference between `reverse!(view(v, lo:hi))` and `reverse!(v, lo, hi)`, except for the unused return value. I may miss smth.

Reply:
Oh, good catch! I didn't know about the `reverse!(v, start, stop)` function, and thought you were referring to `reverse(view(v, lo, hi))`. Thanks for the link. I've removed the redundancy.